components/core/dock.tsx

```diff
@@ -0,0 +1,220 @@
+'use client';
+
+import {
+  motion,
+  MotionValue,
+  useAnimation,
+  useMotionValue,
+  useSpring,
+  useTransform,
+  type SpringOptions,
+} from 'framer-motion';
+import {
+  Children,
+  cloneElement,
+  createContext,
+  useContext,
+  useEffect,
+  useRef,
+} from 'react';
+import { cn } from '@/lib/utils';
+
+/** Constants */
+const DOCK_HEIGHT = 128;
+const DEFAULT_MAGNIFICATION = 80;
+const DEFAULT_DISTANCE = 150;
+/** End of constants */
+
+/** Types */
+type TDock = {
+  className?: string;
+  children: React.ReactNode;
+  spring?: SpringOptions;
+  magnification?: number;
+  distance?: number;
+};
+type TDockItem = {
+  className?: string;
+  children: React.ReactNode;
+};
+type TDockLabel = {
+  className?: string;
+  children: React.ReactNode;
+};
+type TDockIcon = {
+  className?: string;
+  children: React.ReactNode;
+};
+
+type DocContextType = {
+  mouseX: MotionValue;
+  spring: SpringOptions;
+  magnification: number;
+  distance: number;
+};
+type DockProviderProps = {
+  children: React.ReactNode;
+  value: DocContextType;
+};
+/** End of types */
+
+/** Context */
+const DockContext = createContext<DocContextType | undefined>(undefined);
+
+function DockProvider({ children, value }: DockProviderProps) {
+  return <DockContext.Provider value={value}>{children}</DockContext.Provider>;
+}
+/** End of context */
+
+/** Hooks */
+function useDock() {
+  const context = useContext(DockContext);
+  if (!context) {
+    throw new Error('useDock must be used within an DockProvider');
+  }
+  return context;
+}
+/** End of hooks */
+
+function Dock({
+  children,
+  className,
+  spring = { mass: 0.1, stiffness: 150, damping: 12 },
+  magnification = DEFAULT_MAGNIFICATION,
+  distance = DEFAULT_DISTANCE,
+}: TDock) {
+  const mouseX = useMotionValue(Infinity);
+  const isHovered = useMotionValue(0);
+
+  const height = useTransform(
+    isHovered,
+    [0, 1],
+    ['auto', Math.max(DOCK_HEIGHT, magnification + magnification / 2 + 4)]
+  );
+
+  return (
+    <motion.div
+      style={{
+        height: height,
+        scrollbarWidth: 'none',
+      }}
+      className='mx-2 flex max-w-full items-end overflow-x-auto'
```

Comment:
When I hover out of a DockItem on desktop, there's an overflow bug. Removing `overflow-x-auto` seems to fix it.

Reply:
> When I hover out of a DockItem on desktop, there's an overflow bug. Removing `overflow-x-auto` seems to fix it.

I can't remove `overflow-x-auto` because it would disable scrolling on overflow content or smaller screens. I'll look for a workaround.